agda: Remove `--local-interfaces` workaround

ibbem opened the pull request:
## Description of changes

Remove the `--local-interfaces` flag which was a workaround for https://github.com/agda/agda/issues/4613 and https://github.com/agda/agda/issues/4526.
Changelog making this change possible: https://github.com/agda/agda/blob/master/doc/release-notes/2.6.4.2.md
Relevant upstream PR which also fixes #83864: https://github.com/agda/agda/pull/6988

This is currently a draft because I manually changed `pkgs/development/haskell-modules/hackage-packages.nix` to update to the latest Agda release. As far as I know the usual process is to wait until Agda is updated by an automatic script updating `pkgs/development/haskell-modules/hackage-packages.nix`. As soon as the new Agda version is in nixpkgs, I will remove the second commit and then this PR can be merged.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
  I  haven't got `nixpkgs-review` to work (not enough memory).
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ibbem committed:
agda: Remove the --local-interfaces flag

Upstream now provides a library file for the builtin library and ensured
that the existing interface files will be used regardless of whether
--local-interfaces is in effect. Hence, Agda will not try to write to
the Nix store anymore except if the build flags are changed.

ncfavier commented on pkgs/build-support/agda/lib.nix:
```diff
@@ -6,7 +6,7 @@
   * interfaceFile "src/Everything.lagda.tex" == "src/Everything.agdai"
   */
-  interfaceFile = agdaFile: lib.head (builtins.match ''(.*\.)l?agda(\.(md|org|rst|tex|typ))?'' agdaFile) + "agdai";
+  interfaceFile = agdaVersion: agdaFile: "./_build/" + agdaVersion + "/agda/" + lib.head (builtins.match ''./(.*\.)l?agda(\.(md|org|rst|tex|typ))?'' agdaFile) + "agdai";
```
```suggestion
  interfaceFile = agdaVersion: agdaFile: "_build/" + agdaVersion + "/agda/" + lib.head (builtins.match ''(.*\.)l?agda(\.(md|org|rst|tex|typ))?'' agdaFile) + "agdai";
```

ibbem replied:
In `pkgs/development/libraries/agda/1lab/default.nix:25` the assumption of a leading `./` is broken. The leading `./` could be made optional to be more robust I guess. However, I think that means that the everything file (`src/index.lagda.md`) isn't exlcuded as well. I guess in this case this is actually desired.

I think I will add another commit to make the `everythingFile` parameter more robust and add a parameter to disable removal of this file.

ncfavier replied:
Ugh, I always forget that `find -path` compares strings and not paths 🙄.

What's the point of excluding the everything file anyway? Can we just leave it in?

ibbem replied:
Excerpt from `git log`:

> commit 226299e1a2e8dbc9ee8b10042182d8a1f47d7f16
> Author: Alexander Ben Nasrallah <me@abn.sh>
> Date:   Mon Jan 18 20:01:31 2021 +0100
> 
>     agdaPackages.mkDerivation: don't install Everything module
>     
>     The Everthing module is not part of a library and should therefore
>     not be copied to the nix store.
>     
>     This is particularly bad, if the Everything module is defined in
>     an agda library included directory, e.g. consider an agda-lib with
>     
>         include: .
>     
>     and Everything.agda in the project root (.), in which case the
>     Everything module would become part of the library.
>     If multiple such projects are in the dependency tree, the Everything
>     module becomes ambiguous and the build would fail.

However, in the case of 1lab and agdarsec the everything file kind of act as documentation so I'm not sure what should be done there.

ibbem replied:
As a fix I exchanged the `-path` exclusion from `find` by removing the file directly before installing. This should make it more robust in parsing the file path and prevent the builder from continuing in case the exclusion fails.

ncfavier replied:
Thanks, this seems like a sensible fix.

In the case of 1lab, upstream recently abandoned the "index as Everything file" idea, so future updates will have to find another way to generate an Everything file (and in that case we will want it to be removed), so all is good.

ncfavier replied:
Well, except now `cubical` fails to build because the default `Everything.agda` isn't there. We should not fail if `rm` fails; perhaps add `-f`?

ibbem replied:
Oh, right. I missed that.
However, I think it's still a good idea to catch potential errors (e.g., naming changes regarding the interface file). Hence, I decided to move the `rm` into `buildPhase`. This ensures that there will always be an everything file, otherwise the build would fail anyways.

Do you think that's a reasonable compromise?

ncfavier replied:
Yes that's perfect.

ibbem committed:
agda: Don't be too picky about `everythingFile`

The `-path` test of `find` does string comparison, not path comparison.
Hence, the format of `everythingFile` needed to be very specific. Now,
it can be denormalized (e.g. it can contain `/./`) and an error is
emitted if the everything file or its interface file can't be removed.

turion commented:
CC @neosimsim @alexarice

alexarice commented:
Agda 2.6.4.2 has been pulled, so this should presumably wait until 2.6.4.3 has been released

ibbem commented:
Agda 2.6.4.3 has been released and the release is included in `haskell-updates`. I rebased onto `haskell-updates` and removed the manual update commit. Hence, this PR is ready to be merged from my point of view.
Do you have any further change requests or suggestions?

ncfavier commented:
I'll test this PR when the new Agda hits the cache.

ncfavier commented:
@ofborg build agda.tests.allPackages

ncfavier commented on pkgs/build-support/agda/lib.nix:
```diff
@@ -6,7 +6,7 @@
   * interfaceFile "Everything.agda" == "Everything.agdai"
   * interfaceFile "src/Everything.lagda.tex" == "src/Everything.agdai"
   */
```
The comment here needs updating, as well as the tests in `nixos/tests/agda.nix` (but those seem entirely useless, so probably just remove them).

ibbem replied:
Good catch. I have fixed the comment and remove the tests (the `rm` should be test enough).

ncfavier commented:
@nixos test agda
